Initial commit for powervs

mfriesenegger opened the pull request:
The goal of this is to deploy a single HANA on PowerVS instance (no HANA system replication or HA).

arbulu89 commented on powervs/infrastructure.tf:
```diff
@@ -0,0 +1,14 @@
+}
+
+terraform {
+  required_version = ">= 0.12"
```
Change this by `0.13`. This is the version we ship and support

mfriesenegger replied:
Done.

arbulu89 reviewed (commented):
Hey @mfriesenegger ,
It looks pretty good so far. Not much to complain. 
Keep going!

mfriesenegger commented:
Force-push 068c887 includes updates which provision a single PowerVS instance through the predeployment operation.

mfriesenegger commented:
Force-push 88e7251 includes updates which provision a single PowerVS instance successfully installing SAP HANA.  The deployment has been tested using the IBM Cloud PowerVS approved image named Linux-SUSE-SAP-12-4.  

Several additional changes have been made but not included in this PR.

- Added an if statement to workaround_payg_cleanup because paygo does not apply to powervs.
- Currently using NFS from another PowerVS instances for SAP media storage.
- Removed /var/run/reboot-needed test condition because it is not implemented in 12 SP4.
- Updating python in provision.sh because salt may fail on minion_configuration if an older python is installed on 12 SPx systems.

@arbulu89 should these changes be be included in this PR, ignored or fixed outside of this PR?

mfriesenegger commented:
Force-push 3e13164 includes updates not included in 88e7251 to provision a single PowerVS instance successfully installing SAP HANA.

arbulu89 commented on generic_modules/salt_provisioner/main.tf:
```diff
@@ -46,10 +46,6 @@ resource "null_resource" "provision" {
     password    = var.password
     private_key = var.private_key != "" ? var.private_key : ""
-
-    bastion_host        = var.bastion_host
```
We cannot remove this. It is used by the other providers

mfriesenegger replied:
This is temporary as I was learning the automation.  I will be enabling and testing the bastion functionality with powervs in a future PR.

arbulu89 commented on generic_modules/salt_provisioner/main.tf:
```diff
@@ -70,7 +66,7 @@ resource "null_resource" "provision" {
 
   provisioner "remote-exec" {
     inline = [
-      "${var.reboot} && [ -f /var/run/reboot-needed ] && echo \"Rebooting the machine...\" && (nohup sudo sh -c 'systemctl stop sshd;/sbin/reboot' &) && sleep 5",
```
We cannot remove this

mfriesenegger replied:
I removed /var/run/reboot-needed test condition because it is not implemented in 12 SP4.

arbulu89 commented on powervs/main.tf:
```diff
@@ -0,0 +1,104 @@
+}
+
+locals {
+  hana_cluster_vip           = ""
```
We should have the auto-generator if the IP addresses here. But this might come in a next PR

arbulu89 commented on powervs/main.tf:
```diff
@@ -0,0 +1,104 @@
+  additional_packages                 = var.additional_packages
+  public_key                          = var.pi_key_pair_name
+  private_key                         = var.private_key
+  authorized_keys                     = []
```
Why is this empty?

mfriesenegger replied:
IBM Cloud has authorized keys capability for PowerVS which I am currently using.  At this point, I have not needed to test this  authorized_keys variable.

arbulu89 commented on powervs/main.tf:
```diff
@@ -0,0 +1,104 @@
+  reg_additional_modules              = var.reg_additional_modules
+  ha_sap_deployment_repo              = var.ha_sap_deployment_repo
+  additional_packages                 = var.additional_packages
+  public_key                          = var.pi_key_pair_name
```
In the other providers this is name `var.public_key`, why do we have a different name here?

mfriesenegger replied:
pi_key_pair_name is a required input parameter - https://cloud.ibm.com/docs/ibm-cloud-provider-for-terraform?topic=ibm-cloud-provider-for-terraform-power-vsi#power-instance.  I did made this change while learning the automation.  I am open to changing this is a future PR.

arbulu89 commented on powervs/modules/hana_node/main.tf:
```diff
@@ -0,0 +1,51 @@
+}
+
+locals {
+  create_ha_infra             = var.hana_count > 1 && var.common_variables["hana"]["ha_enabled"] ? 1 : 0
```
You might need to format this file to have a better indenting using `terraform fmt` command in this folder

arbulu89 commented on powervs/modules/hana_node/main.tf:
```diff
@@ -0,0 +1,51 @@
+resource "ibm_pi_volume" "ibm_pi_hana_volume"{
+  count                = var.hana_count * local.disks_number
+  pi_volume_size       = local.disks_size[count.index - (local.disks_number * floor(count.index / local.disks_number))]
+  pi_volume_name       = "${terraform.workspace}-${var.name}-volume${count.index + 1}"
```
We shouldn't use `terraform.workspace` here. We should use `deployment_name`

arbulu89 commented on powervs/modules/hana_node/main.tf:
```diff
@@ -0,0 +1,51 @@
+  pi_sys_type           = var.pi_sys_type
+  pi_proc_type          = "shared"
+  pi_image_id           = var.os_image
+  pi_instance_name      = "${terraform.workspace}-${var.name}0${count.index + 1}"
```
We should use `deployment_name` instead of `terraform.workspace`

arbulu89 commented on salt/hana_node/hana_inst_media.sls:
```diff
@@ -9,7 +9,7 @@ hana_inst_directory:
     - mode: "0755"
     - makedirs: True
-  {% if grains.get('provider') in ['libvirt', 'openstack'] %}
+  {% if (grains['provider'] == 'libvirt') or (grains['provider'] == 'powervs') %}
```
Add `powervs` in the list together with `libvirt` and `openstack`

arbulu89 commented on salt/hana_node/mount/powervs.sls:
```diff
@@ -0,0 +1,55 @@
+# Configure physical volumes
+{% for disks in grains['hana_data_disks_configuration']['disks_size'].split(',') -%}
+{% set hostname = grains['hostname'] %}
```
You don't need need to have this `hostname` assigment within the loop

arbulu89 commented on salt/hana_node/mount/powervs.sls:
```diff
@@ -0,0 +1,55 @@
+# Configure physical volumes
+{% for disks in grains['hana_data_disks_configuration']['disks_size'].split(',') -%}
+{% set hostname = grains['hostname'] %}
+{%- set wwns = grains['hana_data_disks_wwn'][hostname].split(',') %}
```
I don't really understand why you get the data with `hostname`. Why are you sending data of other nodes, is this needed?
Maybe you should change the grains creation to only have data of this node

arbulu89 commented on salt/os_setup/init.sls:
```diff
@@ -1,5 +1,7 @@
 include:
+  {% if grains['provider'] == 'libvirt' %}
```
You cannot add this. It is used by the other providers too

arbulu89 reviewed (changes requested):
Hi @mfriesenegger ,

As far as I see the PR needs a lot of work. I couldn't add all the comments.
Some important notes:
- I see that you do changes in the code that is used by the other providers. This is really risky as it changes the behaviour of the others, and we don't want this. You should only include the powervs part to run only when is applied
- The `terraform.workspace` should be only used to create the `deployment_name` variable.

I see 2 options. We work on this PR until fixing everything, or we just merge it to the `powervs` branch and you continue with the work. And we do the the final revision when you have the whole thing finished.

I don't know which want is the best honestly, but I'm sure that if we merged this to `master/develop` it would break many things

arbulu89 reviewed (approved):
@mfriesenegger As talked in person, PR approved, so we will get a final deep review when the branch gets moved to `Master`
